### PR TITLE
Add CPS parser, use both CPS and direct style parsers

### PR DESCRIPTION
**Pull request by harendra-kumar**
CPS is good at scalable applicative/monad/alternative operations and direct style is good at stream fusion providing better performance. We use both via rewrite rules depending on the situation.

### CODE REVIEW

File: src/Streamly/Internal/Data/Zipper.hs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+-- /Internal/
+--
+{-# INLINE release #-}
+release :: Zipper m a -> Zipper m a
+release (Zipper [] _ _ _) = error "Bug: release, no checkpoint exists!"
+release (Zipper (n:cps) xs ys stream) =
+    assert (n <= length xs) $
+            case cps of
+                [] -> assert (n == length xs) $ Zipper [] [] ys stream
+                cp:rest -> Zipper ((cp + n) : rest) xs ys stream
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Zipper.hs:153]: Is this releasing the values held by the checkpoint?

File: src/Streamly/Internal/Data/Zipper.hs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+fromList xs = Zipper [] [] xs K.nil
+
+-- | Add a checkpoint to the Zipper so that we do not release the buffer beyond
+-- the checkpoint.
+--
+-- /Internal/
+--
+{-# INLINE checkpoint #-}
+checkpoint :: Zipper m a -> Zipper m a
+checkpoint (Zipper cps xs ys stream) = Zipper (0:cps) xs ys stream
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Zipper.hs:140]: Can you please elaborate on this?

File: src/Streamly/Internal/Data/Zipper.hs [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+-------------------------------------------------------------------------------
+
+-- | @Zipper checkpoints lefts rights tail@.  The focus is on the first element
+-- of @rights@.  @lefts@ is buffered data on the right of the cursor.  @tail@
+-- is a stream that is used to generate more data if the cursor moves past
+-- @rights@.
+--
+-- @checkpoints@ is a stack of checkpoints. A new checkpoint is created by a
+-- @checkpoint@ operation. A checkpoint consists of a count that tracks how
+-- many elements we have yielded after the checkpoint was taken.  We need this
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Zipper.hs:93]: > A checkpoint consists of a count that tracks how many elements we have yielded after the checkpoint was taken.
I'm unable to follow this statement.

Comment (adithyaov) [inline src/Streamly/Internal/Data/Zipper.hs:93]: "after the checkpoint was taken" == "after the checkpoint was created"?

File: src/Streamly/Internal/Data/Parser/ParserK/Types.hs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+#include "inline.hs"
+
+-- |
+-- Module      : Streamly.Internal.Data.Parser.ParserK.Types
+-- Copyright   : (c) 2020 Composewell Technologies
+-- License     : BSD3
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/Parser/ParserK/Types.hs:11]: The appropriate SPDX identifier that cabal uses is "`BSD-3-Clause`"

File: src/Streamly/Internal/Data/Zipper.hs [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+-- | Add a checkpoint to the Zipper so that we do not release the buffer beyond
+-- the checkpoint.
+--
+-- /Internal/
+--
+{-# INLINE checkpoint #-}
+checkpoint :: Zipper m a -> Zipper m a
+checkpoint (Zipper cps xs ys stream) = Zipper (0:cps) xs ys stream
+
+-- | Release the latest checkpoint, releases any values held by the checkpoint.
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Zipper.hs:142]: It does not necessarily release values held by the checkpoint. Or did I misunderstand what releasing values mean? According to me, releasing values = discarding values from the buffer.

File: src/Streamly/Internal/Data/Zipper.hs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+    where
+
+    {-# INLINE go #-}
+    go st buf !acc =
+        let stop = do
+                pst <- acc
+                r <- try $ extract pst
+                return $ case r of
+                    Left (e :: SomeException) ->
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/Zipper.hs:214]: We should `catch` `ParseError` here.

File: src/Streamly/Internal/Data/Parser/ParserK/Types.hs (hunk 4 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Parser/ParserK/Types.hs:11]: Changed. But we need to change all the old files in the source code as well.

File: src/Streamly/Internal/Data/Zipper.hs (hunk 5 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Zipper.hs:142]: It releases the latest checkpoint, if the values are held by that checkpoint they are released from that checkpoint's perspective but the values may be held by other checkpoints so they will keep holding those. I am adding this note:

```
-- Note that the values may still be held by other checkpoints in the stack of
-- checkpoints.
```

File: src/Streamly/Internal/Data/Zipper.hs (hunk 6 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Zipper.hs:214]: Yes, thanks!

File: src/Streamly/Internal/Data/Zipper.hs [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+                        (nil, Left (displayException e))
+                    Right b -> (nil, Right b)
+            single x = do
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Zipper.hs:217]: Good suggestion. Changed. If you have a good proposal for merging go/gobuf you can try it if the performance remains good we can make the change.

File: src/Streamly/Internal/Data/Zipper.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Zipper.hs:93]: done.

**Pull request merged**